src/refract/JsonValue.cc

```diff
@@ -0,0 +1,499 @@
+
+        options = updateTypeAttributes(e, options);
+
+        if (e.empty()) {
+            auto enumerationsIt = e.attributes().find("enumerations");
+            if (e.attributes().end() != enumerationsIt) {
+                const auto enums = get<const ArrayElement>(enumerationsIt->second.get());
+                assert(enums);
+                if (!enums->empty())
+                    for (const auto& enumEntry : enums->get()) {
+                        assert(enumEntry);
+                        return renderValue(*enumEntry, inherit_flags(options));
+                    }
+            }
+
+            if (options.test(NULLABLE_FLAG)) {
+                return so::Null{};
+            }
+
+            LOG(warning) << "no value found for EnumElement; using `null`";
+            return so::Null{};
+        }
+
+        assert(e.get().value());
+        return renderValue(*e.get().value(), inherit_flags(options));
+    } // namespace
+
+    so::Value renderValue(const NullElement& e, TypeAttributes options)
+    {
+        LOG(debug) << "rendering NullElement to JSON Value";
+
+        return so::Null{};
+    }
+
+    so::Value renderValue(const StringElement& e, TypeAttributes options)
+    {
+        LOG(debug) << "rendering StringElement to JSON Value";
+
+        options = updateTypeAttributes(e, options);
+
+        if (e.empty()) {
+            if (options.test(NULLABLE_FLAG))
+                return so::Null{};
+
+            if (const auto& sampleValue = findSample(e)) {
+                return renderValue(*sampleValue, options);
+            }
+
+            if (const auto& defaultValue = findDefault(e)) {
+                return renderValue(*defaultValue, options);
+            }
+
+            LOG(warning) << "no value for non-nullable empty StringElement; using `\"\"`";
+            return so::String{};
+        }
+
+        return so::String{ e.get().get() };
+    }
+
+    so::Value renderValue(const NumberElement& e, TypeAttributes options)
+    {
+        LOG(debug) << "rendering NumberElement to JSON Value";
+
+        options = updateTypeAttributes(e, options);
+
+        if (e.empty()) {
+            if (options.test(NULLABLE_FLAG))
+                return so::Null{};
+
+            if (const auto& sampleValue = findSample(e)) {
+                return renderValue(*sampleValue, options);
+            }
+
+            if (const auto& defaultValue = findDefault(e)) {
+                return renderValue(*defaultValue, options);
+            }
+
+            LOG(warning) << "no value for non-nullable empty NumberElement; using `0`";
+            return so::Number{};
+        }
+
+        return so::Number{ e.get().get() };
+    }
+
+    so::Value renderValue(const BooleanElement& e, TypeAttributes options)
+    {
+        LOG(debug) << "rendering BooleanElement to JSON Value";
+
+        options = updateTypeAttributes(e, options);
+
+        if (e.empty()) {
+            if (options.test(NULLABLE_FLAG))
```

Comment:
Given the implementation for all primitive types are the same, can we make this re-usable instead of duplicating the business logic on value precedence in each renderValue?